rfcs/2020-11-01-remap-metrics-support.md

```diff
@@ -0,0 +1,191 @@
+if the metric was an `AggregatedHistogram`. Similar problems would occur
+if the data was represented through paths rather than functions.
+
+This functionality is provided by the `Lua` transform, so there is still a way
+forward for the user if they need this functionality.
+
+Note that going with the original proposal does not preclude this, so this could
+still be implemented later as a next stage.
+
+### Allow conversion between metric and log events
+
+Currently the RFC just lays out methods to manipulate either log or metric 
+events. It could also be possible to add functions to convert a log event into
+a metric event and viceversa.
+
+For example:
+
+`. = to_log()`
+
+could take the current metric event and write the fields into the root path to 
+create a log event out of the metric.
+
+To create metric events one function per metric type could be provided such as:
+`make_counter(value)`, `make_gauge(value)`, `make_set([value])`
+
+The other functions provided in this [proposal](#doc-level-Proposal) can then be
+used to fill out the rest of the event.
+
+Note this could still be implemented at a later stage.
+
+### Add functions to enable manipulating metrics for the most likely usecases
+
+If we analyse the most likely scenarios where users would want to manipulate
+the metrics, we could provide functions that cover these scenarios.
+
+For example, one likely scenario would be to sample a stream of gauge values and
+convert this into a sampled distribution. A function could be provided that 
+enables this: `sample_gauge_values([buckets], sample_rate)`.
+
+The complexities involved with providing such functions are that it means
+there is no longer a strict one-to-one correspondance between input and output 
+events. Many input events would be summarised to a single output event.
+
+We would need to think about what would need to be done should 
+`sample_gauge_values` be called twice in the script.
+
+This can be handled by the `Lua` transform, but the script to do this can get
+quite involved. If it could be brought into the `Remap` language and simplified
+so than 80% of the likely usecases are covered, it could be a big win for 
+simplifying users configurations.
+
+Note that this could still be implemented later as a next stage.
+
+## Outstanding Questions
+
+- Should we add an extra option to the Remap transform of event_type to force 
+  the user to specify if their script is for log or metric events and error at
+  load time if they use functions specified for the other type?
+
+## Plan Of Attack
+
+Incremental steps that execute this change. Generally this is in the form of:
+
+- [ ] Adjust the Remap transform to allow it to handle both log and metric.
+- [ ] Implement `remap-lang::Object` for metric events. `remap-lang::Object`
+      may need a new function to `get_metric` to return the metric that the
+      metric events will work with. The other functions will just return an 
+      error.
```

Comment:
Depending on which implementation we go with, if we decide to switch to a path-based solution, the implementation of `trait Object` can use the existing methods to decide what it needs to do with the given paths.

E.g. it would have a match statement on the path, and then decide which data to return (e.g. `.namespace` vs `.kind`, etc.). This would remove the need for a specialized `get_metric` function, or the need to always error on other functions.